[AArch64][SVE2p1] Allow more uses of mask in performActiveLaneMaskCombine

kmclaughlin-arm opened the pull request:
The combine replaces a get_active_lane_mask used by two extract subvectors with
a single paired whilelo intrinsic. When the instruction is used for control
flow in a vector loop, an additional extract of element 0 may introduce
other uses of the intrinsic such as ptest and reinterpret cast, which
is currently not supported.

This patch changes performActiveLaneMaskCombine to count the number of
extract subvectors using the mask instead of the total number of uses,
and returns the concatenated results of get_active_lane_mask.

kmclaughlin-arm committed:
[AArch64][SVE2p1] Allow more uses of mask in performActiveLaneMaskCombine

The combine replaces a get_active_lane_mask used by two extract subvectors with
a single paired whilelo intrinsic. When the instruction is used for control
flow in a vector loop, an additional extract of element 0 may introduce
other uses of the intrinsic such as ptest and reinterpret cast, which
is currently not supported.

This patch changes performActiveLaneMaskCombine to count the number of
extract subvectors using the mask instead of the total number of uses,
and allows other uses by these additional operations.

kmclaughlin-arm committed:
- Tests for get_active_lane_mask with more uses than extract_vector

kmclaughlin-arm committed:
- Return a concat_vector created from the results of whilelo_x2 from performActiveLaneMaskCombine
- Add tests for the 4 extracts case which will use ptest & reinterpret_cast
- Remove changes to canRemovePTestInstr

david-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -18774,7 +18774,7 @@ static SDValue performVecReduceAddCombineWithUADDLP(SDNode *N,
 performActiveLaneMaskCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI,
                              const AArch64Subtarget *ST) {
-  if (DCI.isBeforeLegalize())
+  if (DCI.isBeforeLegalize() && !!DCI.isBeforeLegalizeOps())
```
nit: The `!!` looks a little odd. Is it possible to just use `DCI.isBeforeLegalizeOps()`?

kmclaughlin-arm replied:
This was a mistake, it should be `!DCI.isBeforeLegalizeOps()`

kmclaughlin-arm committed:
- Fix !!DCI.isBeforeLegalizeOps -> !DCI.isBeforeLegalizeOps

kmclaughlin-arm committed:
- Remove !DCI.isBeforeLegalizeOps
- Create multiple concats in performActiveLaneMaskCombine when there are more than 2 extracts

kmclaughlin-arm committed:
- Moved zero check into isKnownMultipleOf & added test to TypeSizeTest.cpp
- Remove unnecessary {} from getNode in performActiveLaneMaskCombine

kmclaughlin-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -18820,32 +18826,39 @@ performActiveLaneMaskCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI,
   DCI.CombineTo(Extracts[0], R.getValue(0));
   DCI.CombineTo(Extracts[1], R.getValue(1));
+  SmallVector<SDValue> Concats = {DAG.getNode(
+      ISD::CONCAT_VECTORS, DL, DoubleExtVT, {R.getValue(0), R.getValue(1)})};
```
No, I've removed them in both places

kmclaughlin-arm committed:
- Fix comment in performActiveLaneMaskCombine
